### PR TITLE
Fix gradgrad issues with BatchNorm and Advanced Indexing (master)

### PR DESCRIPTION
**Pull request by gchanan**
This adds two tests for gradchecks:

1) That backward is reentrant
2) That input grad sizes match input sizez.

1) Uncovers an issue with BatchNorm which is now fixed
2) Uncovers an issue with advanced indexing which is also now fixed.

### CODE REVIEW

**Comment by gchanan** [comment]
@killeent as mentioned in https://github.com/pytorch/pytorch/pull/2268, we should (either in this PR or another) move the resizing logic for advanced indexing to C++.

**Comment by soumith** [comment]
will leave this for @apaszke @killeent review.

File: torch/csrc/autograd/functions/batch_normalization.h [hunk 1]
```diff
@@ -80,8 +76,6 @@ struct BatchNormBackwardBackward : public Function, public BatchNormParams {
 
   virtual void releaseVariables() override;
 
-  at::Tensor save_mean;
```
Comment (apaszke) [inline torch/csrc/autograd/functions/batch_normalization.h]: Huh! Nice catch!

**Review by apaszke — APPROVED** [review]
LGTM
